kotlinx-coroutines-core/common/src/CancellableContinuation.kt

```diff
@@ -198,6 +198,8 @@ public interface CancellableContinuation<in T> : Continuation<T> {
      */
     @ExperimentalCoroutinesApi // since 1.2.0
     public fun resume(value: T, onCancellation: ((cause: Throwable) -> Unit)?)
+
+    public fun <R: T> resume(value: R, onCancellation: ((cause: Throwable, value: R) -> Unit)?)
```

Comment:
One more: whether we want to break source compatibility in most cases or almost none of them. The latter can be achieved with `onCancellation: ((R).(cause: Throwable) -> Unit)?` and nicely supports the pattern `cont.resume(response) { close() }`, but it can also silently change the semantics of existing code.

Reply:
> It's not the selects performance I worry about, but the general applicability

I looked into whether we can gain some performance anyway, and the best I could manage included a very non-idiomatic way of structuring code to avoid one extra allocation per `resume`. Here: https://github.com/Kotlin/kotlinx.coroutines/commit/0e3a633243ce37f625eeb0bc3651951fc82a9b9d